### PR TITLE
Re-implement leak check in terms of universes

### LINKED ISSUE #46989: Function pointer docs may need updating
We may need documentation updated for function pointers when used in enums. I won't pretend I understand half of what's written here, but here's the log from IRC (#rust_beginners) where we debugged this issue:

1. [Rust Playground (`Sharlin`)](https://play.rust-lang.org/?gist=d25cb785bee399efe590293ab2c1a44a&version=stable)
2. [Rust Playground (`spear2`)](https://play.rust-lang.org/?gist=0cfe1a8e458fa50d5223054f7ff7298c&version=stable])

```
Sharlin    :: vermiculus: okay... wtf [1]

vermiculus :: Sharlin: Right!?

Sharlin    :: what makes fns with ref arguments not implement Ord or
              anything

vermiculus :: I 100% need the argument to be mutable, but I can't have it be
              mutable without being a ref, and apparently it's choking on the ref

spear2     :: maybe it has to do with the reference having an implicit lifetime?

vermiculus :: ...Ok the '100%' is not completely true, but it's the cleanest option

Sharlin    :: spear2: indeed it seems to be

spear2     :: [2]

vermiculus :: spear2: now there's a topic I still don't understand half or less

Sharlin    :: this is definitely something that should be documented in the fn docs :D

vermiculus feels useful :D

spear2     :: vermiculus: the way i understand it, every reference has an associated
              lifetime, but most of the time it can be 'inferred' and you don't see it

Sharlin    :: so this is about that higher ranked trait bound thingie

vermiculus :: spear2: that's the extent of my understanding as well; I get tripped up
              trying to 'infer' the lifetime myself (effectively seeing what the
              compiler would see)

Sharlin    :: `Bar(fn(&i32))` is actually `Bar(for<'a> fn(&'a i32))`

Sharlin    :: which means the fn is not an ordinary function pointer but a higher
              something

vermiculus :: spooky

vermiculus :: that makes me wonder if this behavior is intentional or if there's a
              more explicit means to express this idea

vermiculus :: throwing that explicit lifetime parameter is making the rest of the
              compiler go a little nutty -- wanting lifetime parameters for *everything* :(

Sharlin    :: yeah

Sharlin    :: so it goes
```

### LINKED ISSUE #57639: Regression in trait bounds that are redundant with associated type's HRTB
The title is a mouthful, so let me explain and give some motivation. Say I have a trait `MyTrait` with an associated type `MessageType`. This message type must be deserializable (in the serde sense), i.e., `MessageType: for<'a> Deserialize<'a>`. Now suppose I have a generic struct `GenericStruct<T>` where `T: MyTrait`. I want the struct to contain a message, so I put a thing of type `T::MessageType` inside the struct. I would like `GenericStruct` to also be `Deserialize`able now, so I `#[derive(Deserialize)]`.

Concretely, the code

```rust
use serde::Deserialize;

//trait DeserializeOwned: for<'a> Deserialize<'a> {}

trait MyTrait {
    type MessageType: Sized + for<'a> Deserialize<'a>;
    //type MessageType: Sized + DeserializeOwned;
}

#[derive(Deserialize)]
struct GenericStruct<T: MyTrait>(T::MessageType);
```
**compiles in 1.31.1-stable and fails in 1.33.0-nightly** with error message
```
error[E0308]: mismatched types
  |
  | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ one type is more general than the other
  |
  = note: expected type `for<'a> _IMPL_DESERIALIZE_FOR_GenericStruct::_serde::Deserialize<'a>`
             found type `_IMPL_DESERIALIZE_FOR_GenericStruct::_serde::Deserialize<'de>`
```
Furthermore, if the commented lines are uncommented, and the first `type` line in `MyTrait` is commented out, the code now compiles in stable and nightly.

I don't believe this is version-specific behavior in serde. I would like to make a neater minimal testcase, but I don't know of any auto-derivable traits that only take a lifetime as a parameter.

### LINKED ISSUE #58451: ICE related to lifetimes and traits
Works fine one stable. ICE on beta and nightly. I couldn't reduce or generalize this error any more than this. It appears to be triggered by some unknown type; notice the inner `f()` has no argument here, it could be any type that doesn't meet the bounds.

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5c7dadb018deca07a3fd1669bba32035)

```rust
fn f<I>(i: I)
where
    I: IntoIterator,
    I::Item: for<'a> Into<&'a ()>,
{}

fn main() {
    f(&[f()]);
}
```

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR temporarily restores the leak-check, but implemented in terms of universes. This is not because the leak check behavior was necessarily **correct**, but because (a) we may want to have a transition period and because (b) we want to have more breathing room to work through the full implications of handling higher-ranked types correctly. Note that this PR builds atop https://github.com/rust-lang/rust/pull/58056.

Fixes #58451 
Fixes #46989 
Fixes #57639 

r? @aturon 
cc @arielb1, @lqd 

~~Temporary note: I've not finished running `./x.py test` locally -- I'm confident a lot of error messages in tests will need updating. I sort of expect them to revert to the older, (imo) less good error messages, which is mildly unfortunate. There might be a way to preserve the new error messages, not sure.~~

### CODE REVIEW

File: src/librustc/infer/region_constraints/leak_check.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /// bigger set and the data structures are not setup for that. If
+    /// we wind up keeping some form of this check long term, it would
+    /// probably be better to remove the snapshot parameter and to
+    /// refactor the constraint set.
+    pub fn leak_check(
+        &mut self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        overly_polymorphic: bool,
+        placeholder_map: &PlaceholderMap<'tcx>,
+        _snapshot: &CombinedSnapshot<'_, 'tcx>,
```
Comment (aturon) [inline src/librustc/infer/region_constraints/leak_check.rs]: Just to confirm: this parameter is basically acting as a static check that we're in the scope of a snapshot?

**Review by aturon — APPROVED** [review]
I reviewed the new commits here and they LGTM, modulo a sanity check re: the role of snapshots.

File: src/librustc/infer/region_constraints/leak_check.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/infer/region_constraints/leak_check.rs]: Confirm, I should perhaps write that more explicitly as a comment.

**Comment by Mark-Simulacrum** [comment]
Nominating for compiler team approval for beta backport because it fixes several regressions.

**Comment by Mark-Simulacrum** [comment]
@rust-timer build c3b22200e6f5b70eb1f99ae6944d989ae17a458a

**Comment by Mark-Simulacrum** [comment]
@craterbot run start=master#f66e4697ae286985ddefc53c3a047614568458bb end=try#c3b22200e6f5b70eb1f99ae6944d989ae17a458a mode=check-only p=5

**Comment by pnkfelix** [comment]
hey @nikomatsakis do you know off-hand if this will also address #57464 ?

**Comment by nikomatsakis** [comment]
@pnkfelix it does not appear to fix that test case

**Comment by nikomatsakis** [comment]
OK, so I pushed one final commit that reverts a small part of the universes PR, which restores our behavior on the old-lub-glb-object.rs test case to match stable.

**Comment by Centril** [comment]
Regression tree:
- google-testing1-cli-1.0.8+20181011:
> [INFO] [stderr] Error response from daemon: endpoint with name confident_tesla already exists in network none
- tokei-8.0.1
> [INFO] [stderr] Error response from daemon: endpoint with name confident_tesla already exists in network none

Both spurious. We have no actual reported regressions 🎉

**Comment by pnkfelix** [comment]
We discussed this beta-nomination in the [compiler team meeting yesterday](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.2354818.20weekly.20meeting.202019-02-21/near/159076962), as part of the checkin with the WG-traits working group. 

The executive summary is: We think the risk/reward here is in favor of approving the beta backport; but since this only landed in nightly so recently (notably, *after* the meeting where we were discussing it), we need to keep our eye out for problems.

Here is how @aturon summarized things (and a lot of meeting participants agreed with this synopsis):

> i think we should:
> 
> - land in nightly
> - land beta backport
> - start crater run on beta
> 
> basically get as much info about the impact as we can, ASAP. if any of these reveal problems, *then* we can delay the release

therefore, marking as beta-accepted.

**Pull request merged**